### PR TITLE
Bug with renewing ot long running tasks fixed

### PR DESCRIPTION
**Pull request by ddobric**
Due the issue with SB time precision, this check-in contains a fix, which renews the lock in regular interval of 30 sec. It is recommended to make this interval configurable in dependence of the lock time defined on the entity.

### CODE REVIEW

File: Framework/TaskActivityDispatcher.cs [hunk 1]
```diff
@@ -152,45 +152,45 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
...
-                }
-
-                DateTime renewAt = message.LockedUntilUtc.Subtract(TimeSpan.FromSeconds(30));
-
-                // service bus clock sku can really mess us up so just always renew every 30 secs regardless of 
-                // what the message.LockedUntilUtc says. if the sku is negative then in the worst case we will be
-                // renewing every 5 secs
-                //
-                renewAt = AdjustRenewAt(renewAt);
+                int renewInterval = 30000;
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: We need to add the renewInterval as a configurable setting to TaskActivityDispatcherSettings. It should default to 30 secs.

File: Framework/TaskActivityDispatcher.cs [hunk 2]
```diff
@@ -152,45 +152,45 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
...
 
                 while (!cancellationToken.IsCancellationRequested)
                 {
-                    await Task.Delay(TimeSpan.FromSeconds(5));
-
-                    if (DateTime.UtcNow >= renewAt)
+                    await Task.Delay(TimeSpan.FromSeconds(30));
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: This should actually use the renewInterval defined above.

File: Framework/TaskActivityDispatcher.cs [hunk 3]
```diff
@@ -152,45 +152,45 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
...
-                            // might have been completed
-                            TraceHelper.TraceException(TraceEventType.Information, exception,
-                                "Failed to renew lock for message {0}", message.MessageId);
-                            break;
+                            try
+                            {
+                                TraceHelper.Trace(TraceEventType.Information, "Renewing lock for message id {0}",
+                                    message.MessageId);
+
+                                message.RenewLock();
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: Btw can we switch to await message.RenewLockAsync()? 

Not sure why we were not doing that already :)

File: Framework/TaskActivityDispatcher.cs [hunk 4]
```diff
@@ -152,45 +152,45 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
...
+
+                                TraceHelper.Trace(TraceEventType.Information, "Next renew for message id '{0}' at '{1}'",
+                                    message.MessageId, DateTime.Now.AddMilliseconds(renewInterval));
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: We want to normalize to UTC time everywhere. Can this be changed to DateTime.UtcNow.AddMilliseconds() etc?

File: Framework/TaskActivityDispatcher.cs [hunk 5]
```diff
@@ -152,45 +152,45 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
...
+        /// LockDuration, which is usually 1 minute by default.
+        /// It would be good to make default renew interval configurable.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="cancellationToken"></param>
         async void RenewUntil(BrokeredMessage message, CancellationToken cancellationToken)
         {
             try
             {
-                if (message.LockedUntilUtc < DateTime.UtcNow)
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: Might be worth it to still keep this check around.

File: Framework/TaskActivityDispatcher.cs [hunk 6]
```diff
@@ -202,6 +202,8 @@ public sealed class TaskActivityDispatcher : DispatcherBase<BrokeredMessage>
             }
         }
 
+
+
         DateTime AdjustRenewAt(DateTime renewAt)
```
Comment (affandar) [inline Framework/TaskActivityDispatcher.cs]: I think this method can be deleted now that it is not used.

**Comment by affandar** [comment]
Damir, thanks for sending this PR. It greatly simplifies the lock renewal code and I would love to accept the change after you go through the comments I've given above.

Also please make sure that all the unit tests are passing with this change.

**Comment by ddobric** [comment]
What is the status of this?

**Comment by simonporter** [comment]
Closing old pull requests.
Please do submit new requests if the changes are applicable.

**Pull request closed without merging**